dev/build-docker-image-matrix.py

```diff
@@ -121,26 +124,23 @@ def tag_latest_ubuntu_with_flwr_version(image: BaseImage) -> List[str]:
         return [image.tag]
 
 
-if __name__ == "__main__":
-    arg_parser = argparse.ArgumentParser(
-        description="Generate Github Docker workflow matrix"
-    )
-    arg_parser.add_argument("--flwr-version", type=str, required=True)
-    args = arg_parser.parse_args()
-
-    flwr_version = args.flwr_version
+def build_complete_matrix(flwr_version: str):
+    """Generates a matrix comprising Ubuntu and Alpine images. For Alpine, the matrix
+    includes only the latest supported Python version, whereas for Ubuntu, it includes all
+    supported Python versions.
+    """
```

Comment:
We don't build Alpine images for the supported Python versions because we want to keep the installation of dependencies simple. Alpine uses `musl` instead of `glibc`, and most ML frameworks don't provide binaries that are compiled for `musl`, which means they must be compiled when building the image. This process can take more than 2 hours depending on the framework and hardware, making it a time-consuming and resource-intensive task.

As a result, we only use the Alpine base image for building images where we think the user doesn't need to install any dependencies, such as the SuperLink and SuperNode image. For images where the user might need to install dependencies, we provide Ubuntu images for all supported Python versions.

The `build_complete_matrix` function generates the matrix when a new `flwr` version is released. For testing we will use the simplified matrix (`build_ubuntu_python_latest_matrix`) that only builds Ubuntu images with the latest supported Python version because building Ubuntu images is faster and cheaper that Alpine images.

The Docker workflows (`release`, `nightly`, `main` aka. `unstable`) will always build the images for `amd64` and `arm64`. However, for the Docker e2e test workflow, we agreed to only build and test the `amd64` version.

Later, we could also test Alpine images and `arm64` architecture, but for now, we just want a simple test to ensure the cron job doesn't produce a broken image. https://github.com/adap/flower/pull/4092#pullrequestreview-2296421798

Reply:
@Robert-Steiner Got it. This is a very clear explanation - let's document it somewhere (e.g. in `build-docker-image-matrix.py`?).

Reply:
I added the documentation in https://github.com/adap/flower/pull/4422